File: 백준/Platinum/1102. 발전소/발전소.py
import sys


def input():
    return sys.stdin.readline().rstrip()


def getMinWeight(visit: int, nextNode: int) -> int:
    if dpW[nextNode][visit] != -1:
        return dpW[nextNode][visit]

    dpW[nextNode][visit] = min([graph[curNode][nextNode] for curNode in range(N)
                                if visit & 1 << curNode])
    return dpW[nextNode][visit]


def getMinCost(visit: int, numLeft: int) -> int:
    if numLeft <= 0:
        return 0

    if visit == 0:
        return -1

    if dpC[visit] != INF:
        return dpC[visit]

    ret = INF
    for nextNode in range(N):
        if ~visit & (1 << nextNode):
            val = getMinWeight(visit, nextNode) + getMinCost(visit | (1 << nextNode), numLeft - 1)
            if val < ret:
                ret = val

    dpC[visit] = min([getMinWeight(visit, nextNode) + getMinCost(visit | 1 << nextNode, numLeft - 1)
                      for nextNode in range(N) if not visit & 1 << nextNode])
    return dpC[visit]


def solution() -> int:
    return getMinCost(initBit, P - cntY)


if __name__ == '__main__':
    N = int(input())
    graph = [list(map(int, input().split())) for _ in range(N)]

    INF = int(1e8)

    dpC = [INF for c in range(1 << N)]
    dpW = [[-1 for c in range(1 << N)] for r in range(N)]

    initBit = 0
    s = input()
    cntY = 0
    for i, v in enumerate(s):
        if v == 'Y':
            initBit |= 1 << i
            cntY += 1

    P = int(input())

    print(solution())

File: 백준/Platinum/1102. 발전소/발전소.cc
#include <iostream>
#include <algorithm>
using namespace std;
#define INF static_cast<int>(1e5)


const int MAX_N = 16;
int graph[MAX_N][MAX_N], dpW[MAX_N][1 << MAX_N], dpC[1 << MAX_N];

int N, P;


int getMinWeight(int fixed, int nextNode) {
    int &ret = dpW[nextNode][fixed];
    if (ret != INF) {
        return ret;
    }
    for (int curNode = 0; curNode < N; curNode++) {
        if (fixed & 1 << curNode) {
            ret = min(ret, graph[curNode][nextNode]);
        }
    }
    return ret;
}


int getMinCost(int fixed, int cntFixed) {
    if (cntFixed >= P) {
        return 0;
    }
    if (fixed == 0) {
        return -1;
    }
    int &ret = dpC[fixed];
    if (ret != INF) {
        return ret;
    }
    for (int nextNode = 0; nextNode < N; nextNode++) {
        if (!(fixed & 1 << nextNode)) {
            ret = min(ret, getMinWeight(fixed, nextNode) + getMinCost(fixed | 1 << nextNode, cntFixed + 1));
        }
    }
    return ret;
}


int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    fill(dpC, dpC + (1 << MAX_N), INF);
    fill(&dpW[0][0], &dpW[0][0] + MAX_N * (1 << MAX_N), INF);

    cin >> N;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> graph[i][j];
        }
    }
    string s; cin >> s;
    int initBit = 0, cntY = 0;
    for (int i = 0; i < s.size(); i++) {
        if (s[i] == 'Y') {
            initBit |= 1 << i;
            cntY++;
        }
    }
    cin >> P;
    cout << getMinCost(initBit, cntY);
    return 0;
}